implement apis based on ServiceTemplate

niemannd opened the pull request:
⚠️ This PR includes #120 

This PR contains the api implementations based on the ServiceTemplate. 
IndexHandler to handle Index and the Settings Endpoints
DocumentHandler to handle Document operations
InstanceHandler to handle Instance-wide endpoints like Stats

DocumentHandler is implemented as a generic to support a user-provided model class.
This way we dont have to return a String.

eskombro reviewed (commented):
I started reviewing this PR. The work is huge! Thanks! I added a few specific comments/questions in the process.

I wanted to discuss a few things for now:

1. Is there a possibility to implement this API structure without forcing the user to necessarily create a ServiceTemplate by himself? In the ideal world, the getting started that we have on the README should stay as simple as it is now. Of course if something new or an additional step is needed, it can be added. But we always prioritize the simplicity of use. For example, the level of abstraction that we added with the JsonHandler let the user customize their project by using a specific Json library, but we don't force him to make this choice by adding an additional step: we have a default choice, and we let someone change it if this is what they want. I would like to see this same principle or idea applied to ServiceTemplates: the idea would be that the user only **needs** to setup basic configuration (host, API Key...) and they should be able to start using the Java SDK without **having** to configure an HTTP client, a Json Handler etc. But having this implementation should give them an additional choice. This choice should be optional, but not mandatory.

Can you please add here in a comment how the simplest getting started could be (or the minimum steps user needs to add documents and perform a search)?

2. This is an implementation that is completely independent of a lot of code present at `com/meilisearch/sdk` folder. This is only being used by the integration tests, which are basically not using ServiceTemplates at all. Is this the reason why this files are still there? I'm talking about:

- Documents
- IndexesHandler
- Settings
- SettingsHandler
- Updates
- UpdateStatus 

3. From your POV, should `Dumps` and `DumpsHandler` be a part of `InstanceHandler`?

niemannd commented:
> Is there a possibility to implement this API structure without forcing the user to necessarily create a ServiceTemplate by himself? 

Already done, i just need to create a PR when the three i have open are merged 😄.

> This is an implementation that is completely independent of a lot of code present at com/meilisearch/sdk folder. 

Yeah, i know 😢 . The reason is that every time i tried to refactor the existing codebase everything else imploded. 
So to not send you a single PR with 20000 LoC changes I implemented most of the apis independent from the current code. 
The three PRs i opened are basically the non breaking part of the api. 

The next PR will put it all together and probably break backwards compatibility.
Rewrote [the Client](https://github.com/niemannd/meilisearch-java-forked/blob/feat/config-client-builder/src/main/java/com/meilisearch/sdk/Client.java), [the Config](https://github.com/niemannd/meilisearch-java-forked/blob/feat/config-client-builder/src/main/java/com/meilisearch/sdk/Config.java) and [added a ClientBuilder](https://github.com/niemannd/meilisearch-java-forked/blob/feat/config-client-builder/src/main/java/com/meilisearch/sdk/ClientBuilder.java) 

> From your POV, should Dumps and DumpsHandler be a part of InstanceHandler?

Looks like that [one commit slipped](https://github.com/niemannd/meilisearch-java-forked/commit/64f136d9285a737f5fd5f8b27b7e6b28d761ef57) throu. 🤦‍♂️ I'll fix that.

niemannd commented on src/main/java/com/meilisearch/sdk/api/documents/DocumentHandler.java:
```diff
@@ -0,0 +1,209 @@
+package com.meilisearch.sdk.api.documents;
```
Yeah, looks like i missed that. Lets do this in another PR later. This PR is already complicated enough :D

eskombro replied:
> Yeah, looks like i missed that. Lets do this in another PR later. This PR is already complicated enough :D

Sure, this is why I opened the issue!

niemannd commented on src/main/java/com/meilisearch/sdk/MeiliSearchHttpRequest.java:
```diff
@@ -24,18 +27,20 @@
 	public MeiliSearchHttpRequest(AbstractHttpClient client, RequestFactory factory) {
 		this.client = client;
 		this.factory = factory;
+		this.jsonHandler = new GsonJsonHandler();
```
That was part of the errorhandling rewrite. You already merged that one. 
Honestly, i just needed a deserializer for this class after rewriting the error handling. As MeiliSearchHttpRequest used Gson before, i just put the GsonJsonHandler there 😆

niemannd commented on src/main/java/com/meilisearch/sdk/api/documents/DocumentHandler.java:
```diff
@@ -0,0 +1,209 @@
+	 * @throws MeiliSearchRuntimeException in case something went wrong (http error, json exceptions, etc)
+	 */
+	public List<T> getDocuments() throws MeiliSearchRuntimeException {
+		return getDocuments(20);
```
Yeah, I'll fix that.

eskombro replied:
Thanks :)

eskombro commented:
> > Is there a possibility to implement this API structure without forcing the user to necessarily create a ServiceTemplate by himself?
> 
> Already done, i just need to create a PR when the three i have open are merged 😄.

Ok, great! I am asking if this is the way you want to go in the end, because this is one of the biggest priorities and concerns for MeiliSearch tools: as simple as possible to use, and require as few configuration as possible to get started :)

> > This is an implementation that is completely independent of a lot of code present at com/meilisearch/sdk folder.
> 
> Yeah, i know 😢 . The reason is that every time i tried to refactor the existing codebase everything else imploded.
> So to not send you a single PR with 20000 LoC changes I implemented most of the apis independent from the current code.
> The three PRs i opened are basically the non breaking part of the API.

That's perfect, I think that this is indeed a good way to implement ServiceTemplate while avoiding breaking changes while possible, at least until the implementation of the APIS is complete. So great 👌🏽 

I also know that you prefer to use Unit tests with mock server. This is fine, but the problem I see with this right now is that MeilISearch is not supposed to be stable for some time, and every one of this changes introduces not only maintenance on the tests, but also the risk that we miss something because MeilISearch response is not longer the same than the mocked server. So I still think we need to add (not necessarily now) integration tests. Maybe when we introduce breaking changes, the Integration test should be changed to test the ServiceTemplate API simple mentation, but for now that are doing the job to keep track of breaking changes, so that is great.

Thanks for your answer!

niemannd commented:
> So I still think we need to add (not necessarily now) integration tests. 
Most of the heavy stuff is already done: https://github.com/niemannd/meilisearch-java-forked/tree/master/src/test/java/com/meilisearch/integration
I think the APIs inside InstanceHandler and KeysHandler are not integraiton tested, but i dont think they are _that_ critical.

eskombro commented:
> > So I still think we need to add (not necessarily now) integration tests.
> 
> Most of the heavy stuff is already done: https://github.com/niemannd/meilisearch-java-forked/tree/master/src/test/java/com/meilisearch/integration
> I think the APIs inside InstanceHandler and KeysHandler are not integraiton tested, but i dont think they are _that_ critical.

Amazing!

eskombro commented:
Oh @niemannd I see that you solved some of the "conversations" that were open on the review, but maybe you missed a few because they were collapsed in this Github view (at leasst they are on my GH interface)

niemannd commented:
The requested changes are done, also i rebased onto master, so i think this is ready to merge.

eskombro commented:
bors try

eskombro commented:
Thanks @niemannd !

Merging 🎉 

Bors merge
